EIPS/eip-00.md

```diff
@@ -0,0 +1,831 @@
+They are assigned their own sequential `transactionIndex`, and their `transactionHash` is defined as
+(`AA_TX_TYPE transaction hash + 1`).
+
+All block-related RPC methods, like `eth_getBlockByHash` and `eth_getBlockByNumber`, must include these virtual
+transactions as part of the `transactions` field and include validation in the block transaction count.
+
+All transaction-related RPC methods, like `eth_getTransactionByHash` and `eth_getTransactionReceipt`, must
+accept the virtual transaction hash as input and return the details calculated as if the validation was a
+separate transaction.
+
+There is a number of behaviours that define transaction-wide effects in Ethereum.
+This list includes, but is not limited to:
+
+* Tracking `accessed_addresses`
+* [EIP-1283](./eip-1283) Gas metering for SSTORE
+* [EIP-1153](./eip-1153) Transient storage opcodes
+
+Any such behaviour has separate effects in the "Validation Virtual Transaction" and "Execution Transaction".
+
+Gas refunds are issued at the end of the entire transaction only.
+
+### Transaction validity time range parameters
+
+The `Paymaster validation frame` and the `Sender validation frame` each return values `validUntil` and `validAfter`.
+If the transaction is initiated by an EOA, these fields may be encoded into unused bits of the `nonce`.
+
+These values allow the `sender` and `paymaster` contracts to specify
+a time range for the blocks the transaction will be valid for.
+
+Transaction cannot be included in a block outside of this time range.
+If included, such a block is considered invalid.
+
+Passing `validUntil = 0` and `validAfter = 0` disables the check.
+
+### Calculation of Transaction Type AA_TX_TYPE hash
+
+```
+
+keccak256(AA_TX_TYPE || 0x00 || rlp(transaction_payload)
+
+```
+
+Note that the `chainId` and `accessList` parameters are included in the transaction hash calculation but are not
+available on-chain as part of the `TransactionType4` struct.
+
+In order to calculate the transaction hash that will be used during the signing of the transaction and validation of
+the transaction signature by the `sender`, the value of the `signature` parameter is considered to be an empty
+byte array.
+
+### Accepting EOA account as `sender` to achieve native gas abstraction
+
+In case the `sender` address does not have any code deployed and the `deployerData` length is zero,
+interpret the `signature` parameter as `(y_parity, r, s)` and the `nonce` parameter
+as `(validUntil, validAfter, nonce)`.
+Replace the sender validation frame with default ECDSA signature validation.
+Also check the block timestamp is within the `[validUntil, validAfter]` range.
+
+The base transaction gas cost, in this case, is increased by `AA_ECRECOVER_COST`.
+
+The `callData` parameter in this case is interpreted as following:
+
+```
+
+target || value || data
+
+```
+
+### Execution layer block validation
+
+When validating a block, the validity conditions for a block are extended as follows:
```

Comment:
The validation rules here imply that the block will basically have the following structure:

```
<AA chunk>
<some non-zero number of EOA txs>
<AA chunk>
<some non-zero number of EOA txs>
```

Where each `<AA chunk>` is a set of `N` `AA_TX_TYPE` transactions, where there are strictly firsly N validations and then N executions. While it is definitely fine in a general case, if a builder wants to promise atomicity of AA transactions (i.e. each validation must be necessarily followed by it execution), it would mean that the block is invalid by the rules above (because a valid `<AA chunk>` is _strictly_ of form `Val1, Val2, Ex1, Ex2`, i.e. `Val1, Ex1, Val2, Ex2` are not allowed). So if a builder wanted to have atomic AA transactions, while keeping its block valid, it would have to separate these small `<AA chunks>` s by an EOA transaction. 

Builders promising better UX for users is a common service provided by for instance Flashbots. So IMHO it is important to keep more options available.

Reply:
Right. Recently we've been discussing another reason this is needed, to mitigate a certain DoS attack against builders. We contemplated a NOP transaction (basically a separator transaction that is smaller than an EOA one), or a header transaction that specifies the size of the next AA chunk.  We're thinking of going with the latter - see discussion in https://github.com/eth-infinitism/EIPs/pull/5  You're welcome to weigh in on it.

Reply:
Note that a builder can skip bundling, and create a block of `val1 ex1 <sep> val2 ex2`
However, this voids the tx isolation.
That is, this bundler must verify that val2 wasn't invalidated by ex1. This can easily be an attack vector.

Reply:
Right, and that's why we propose bundling as a general strategy.  But if, for whatever reason, the builder wants to separate bundles (or even do bundlers of 1), we should support it.  We saw that it's necessary for mitigating unused-gas attacks.